File: app/src/main/java/com/mikepenz/fastadapter/app/items/ImageItem.java
package com.mikepenz.fastadapter.app.items;

import android.content.Context;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewPropertyAnimator;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.app.R;
import com.mikepenz.fastadapter.utils.EventHookUtil;
import com.mikepenz.fastadapter.items.AbstractItem;
import com.mikepenz.fastadapter.listeners.ClickEventHook;
import com.mikepenz.iconics.view.IconicsImageView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by mikepenz on 28.12.15.
 */
public class ImageItem extends AbstractItem<ImageItem, ImageItem.ViewHolder> {

    public String mImageUrl;
    public String mName;
    public String mDescription;
    public boolean mStarred = false;

    public ImageItem withImage(String imageUrl) {
        this.mImageUrl = imageUrl;
        return this;
    }

    public ImageItem withName(String name) {
        this.mName = name;
        return this;
    }

    public ImageItem withDescription(String description) {
        this.mDescription = description;
        return this;
    }

    public ImageItem withStarred(boolean starred) {
        this.mStarred = starred;
        return this;
    }

    /**
     * defines the type defining this item. must be unique. preferably an id
     *
     * @return the type
     */
    @Override
    public int getType() {
        return R.id.fastadapter_image_item_id;
    }

    /**
     * defines the layout which will be used for this item in the list
     *
     * @return the layout for this item
     */
    @Override
    public int getLayoutRes() {
        return R.layout.image_item;
    }

    /**
     * binds the data of this item onto the viewHolder
     *
     * @param viewHolder the viewHolder of this item
     */
    @Override
    public void bindView(ViewHolder viewHolder, List<Object> payloads) {
        super.bindView(viewHolder, payloads);

        //get the context
        Context ctx = viewHolder.itemView.getContext();

        //define our data for the view
        viewHolder.imageName.setText(mName);
        viewHolder.imageDescription.setText(mDescription);
        viewHolder.imageView.setImageBitmap(null);

        //we pre-style our heart :D
        style(viewHolder.imageLovedOn, mStarred ? 1 : 0);
        style(viewHolder.imageLovedOff, mStarred ? 0 : 1);

        //load glide
        Glide.with(ctx).load(mImageUrl).animate(R.anim.alpha_on).into(viewHolder.imageView);
    }

    @Override
    public void unbindView(ViewHolder holder) {
        super.unbindView(holder);
        Glide.clear(holder.imageView);
        holder.imageView.setImageDrawable(null);
    }

    @Override
    public ViewHolder getViewHolder(View v) {
        return new ViewHolder(v);
    }

    /**
     * helper method to style the heart view
     *
     * @param view
     * @param value
     */
    private void style(View view, int value) {
        view.setScaleX(value);
        view.setScaleY(value);
        view.setAlpha(value);
    }

    /**
     * helper method to animate the heart view
     *
     * @param imageLovedOn
     * @param imageLovedOff
     * @param on
     */
    public void animateHeart(View imageLovedOn, View imageLovedOff, boolean on) {
        imageLovedOn.setVisibility(View.VISIBLE);
        imageLovedOff.setVisibility(View.VISIBLE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB_MR1) {
            viewPropertyStartCompat(imageLovedOff.animate().scaleX(on ? 0 : 1).scaleY(on ? 0 : 1).alpha(on ? 0 : 1));
            viewPropertyStartCompat(imageLovedOn.animate().scaleX(on ? 1 : 0).scaleY(on ? 1 : 0).alpha(on ? 1 : 0));
        }
    }

    /**
     * helper method for the animator on APIs < 14
     *
     * @param animator
     */
    public static void viewPropertyStartCompat(ViewPropertyAnimator animator) {
        if (Build.VERSION.SDK_INT >= 14) {
            animator.start();
        }
    }

    /**
     * our ViewHolder
     */
    public static class ViewHolder extends RecyclerView.ViewHolder {
        protected View view;
        @BindView(R.id.item_image_img)
        protected ImageView imageView;
        @BindView(R.id.item_image_name)
        protected TextView imageName;
        @BindView(R.id.item_image_description)
        protected TextView imageDescription;
        @BindView(R.id.item_image_loved_container)
        public RelativeLayout imageLovedContainer;
        @BindView(R.id.item_image_loved_yes)
        protected IconicsImageView imageLovedOn;
        @BindView(R.id.item_image_loved_no)
        protected IconicsImageView imageLovedOff;

        public ViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
            this.view = view;

            //optimization to preset the correct height for our device
            int columns = view.getContext().getResources().getInteger(R.integer.wall_splash_columns);
            int screenWidth = view.getContext().getResources().getDisplayMetrics().widthPixels;
            int finalHeight = (int) (screenWidth / 1.5);
            imageView.setMinimumHeight(finalHeight / columns);
            imageView.setMaxHeight(finalHeight / columns);
            imageView.setAdjustViewBounds(false);
            //set height as layoutParameter too
            FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) imageView.getLayoutParams();
            lp.height = finalHeight / columns;
            imageView.setLayoutParams(lp);
        }
    }

    public static class ImageItemHeartClickEvent extends ClickEventHook<ImageItem> {
        @Nullable
        @Override
        public List<View> onBindMany(@NonNull RecyclerView.ViewHolder viewHolder) {
            if (viewHolder instanceof ImageItem.ViewHolder) {
                return EventHookUtil.toList(((ViewHolder) viewHolder).imageLovedContainer);
            }
            return super.onBindMany(viewHolder);
        }

        @Override
        public void onClick(View v, int position, FastAdapter<ImageItem> fastAdapter, ImageItem item) {
            item.withStarred(!item.mStarred);
            //we animate the heart
            item.animateHeart(((ViewGroup) v).getChildAt(0), ((ViewGroup) v).getChildAt(1), item.mStarred);

            //we display the info about the click
            Toast.makeText(v.getContext(), item.mImageUrl + " - " + item.mStarred, Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/mikepenz/fastadapter/app/items/ImageItem.kt
package com.mikepenz.fastadapter.app.items

import android.os.Build
import android.view.View
import android.view.ViewGroup
import android.view.ViewPropertyAnimator
import android.widget.*
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.app.R
import com.mikepenz.fastadapter.items.AbstractItem
import com.mikepenz.fastadapter.listeners.ClickEventHook
import com.mikepenz.iconics.view.IconicsImageView

/**
 * Created by mikepenz on 28.12.15.
 */
class ImageItem : AbstractItem<ImageItem.ViewHolder>() {

    var mImageUrl: String? = null
    var mName: String? = null
    var mDescription: String? = null
    var mStarred = false

    /**
     * defines the type defining this item. must be unique. preferably an id
     *
     * @return the type
     */
    override val type: Int
        get() = R.id.fastadapter_image_item_id

    /**
     * defines the layout which will be used for this item in the list
     *
     * @return the layout for this item
     */
    override val layoutRes: Int
        get() = R.layout.image_item

    fun withImage(imageUrl: String): ImageItem {
        this.mImageUrl = imageUrl
        return this
    }

    fun withName(name: String): ImageItem {
        this.mName = name
        return this
    }

    fun withDescription(description: String): ImageItem {
        this.mDescription = description
        return this
    }

    fun withStarred(starred: Boolean): ImageItem {
        this.mStarred = starred
        return this
    }

    fun withIdentifier(identifier: Long): ImageItem {
        this.identifier = identifier
        return this
    }

    /**
     * binds the data of this item onto the viewHolder
     *
     * @param holder the viewHolder of this item
     */
    override fun bindView(holder: ViewHolder, payloads: List<Any>) {
        super.bindView(holder, payloads)

        //get the context
        val ctx = holder.itemView.context

        //define our data for the view
        holder.imageName.text = mName
        holder.imageDescription.text = mDescription
        holder.imageView.setImageBitmap(null)

        //we pre-style our heart :D
        style(holder.imageLovedOn, if (mStarred) 1 else 0)
        style(holder.imageLovedOff, if (mStarred) 0 else 1)

        //load glide
        Glide.with(ctx).load(mImageUrl).animate(R.anim.alpha_on).into(holder.imageView)
    }

    override fun unbindView(holder: ViewHolder) {
        super.unbindView(holder)
        Glide.clear(holder.imageView)
        holder.imageView.setImageDrawable(null)
    }

    override fun getViewHolder(v: View): ViewHolder {
        return ViewHolder(v)
    }

    /**
     * helper method to style the heart view
     *
     * @param view
     * @param value
     */
    private fun style(view: View, value: Int) {
        view.scaleX = value.toFloat()
        view.scaleY = value.toFloat()
        view.alpha = value.toFloat()
    }

    /**
     * helper method to animate the heart view
     *
     * @param imageLovedOn
     * @param imageLovedOff
     * @param on
     */
    fun animateHeart(imageLovedOn: View, imageLovedOff: View, on: Boolean) {
        imageLovedOn.visibility = View.VISIBLE
        imageLovedOff.visibility = View.VISIBLE

        viewPropertyStartCompat(imageLovedOff.animate().scaleX((if (on) 0 else 1).toFloat()).scaleY((if (on) 0 else 1).toFloat()).alpha((if (on) 0 else 1).toFloat()))
        viewPropertyStartCompat(imageLovedOn.animate().scaleX((if (on) 1 else 0).toFloat()).scaleY((if (on) 1 else 0).toFloat()).alpha((if (on) 1 else 0).toFloat()))
    }

    /**
     * our ViewHolder
     */
    class ViewHolder(var view: View) : RecyclerView.ViewHolder(view) {
        var imageView: ImageView = view.findViewById(R.id.item_image_img)
        var imageName: TextView = view.findViewById(R.id.item_image_name)
        var imageDescription: TextView = view.findViewById(R.id.item_image_description)
        var imageLovedContainer: RelativeLayout = view.findViewById(R.id.item_image_loved_container)
        var imageLovedOn: IconicsImageView = view.findViewById(R.id.item_image_loved_yes)
        var imageLovedOff: IconicsImageView = view.findViewById(R.id.item_image_loved_no)

        init {
            //optimization to preset the correct height for our device
            val columns = view.context.resources.getInteger(R.integer.wall_splash_columns)
            val screenWidth = view.context.resources.displayMetrics.widthPixels
            val finalHeight = (screenWidth / 1.5).toInt()
            imageView.minimumHeight = finalHeight / columns
            imageView.maxHeight = finalHeight / columns
            imageView.adjustViewBounds = false
            //set height as layoutParameter too
            val lp = imageView.layoutParams as FrameLayout.LayoutParams
            lp.height = finalHeight / columns
            imageView.layoutParams = lp
        }
    }

    class ImageItemHeartClickEvent : ClickEventHook<ImageItem>() {
        override fun onBindMany(viewHolder: RecyclerView.ViewHolder): List<View>? {
            if (viewHolder is ViewHolder) {
                return listOf(viewHolder.imageLovedContainer)
            }
            return super.onBindMany(viewHolder)
        }

        override fun onClick(v: View, position: Int, fastAdapter: FastAdapter<ImageItem>, item: ImageItem) {
            item.withStarred(!item.mStarred)
            //we animate the heart
            item.animateHeart((v as ViewGroup).getChildAt(0), v.getChildAt(1), item.mStarred)

            //we display the info about the click
            Toast.makeText(v.context, item.mImageUrl + " - " + item.mStarred, Toast.LENGTH_SHORT).show()
        }
    }

    companion object {

        /**
         * helper method for the animator on APIs < 14
         *
         * @param animator
         */
        fun viewPropertyStartCompat(animator: ViewPropertyAnimator) {
            if (Build.VERSION.SDK_INT >= 14) {
                animator.start()
            }
        }
    }
}
